src/lexer.mll

```diff
@@ -271,8 +274,80 @@ let raw_markup_target =
 let language_tag_char =
   ['a'-'z' 'A'-'Z' '0'-'9' '_' '-' ]
 
+rule reference_paren_content input start depth_paren depth_curly buffer = parse
+  | '('
+    {
+      buffer_add_lexeme buffer lexbuf ;
+      reference_paren_content input start (depth_paren + 1) depth_curly buffer
+        lexbuf }
+  | '{'
+    {
+      buffer_add_lexeme buffer lexbuf ;
+       reference_paren_content input start depth_paren (depth_curly + 1) buffer
+        lexbuf }
+  | ')'
+    {
+      buffer_add_lexeme buffer lexbuf ;
+      if depth_paren = 0 then false
+      else
+        ( reference_paren_content input start (depth_paren - 1) depth_curly
+            buffer lexbuf ) }
+  | '}'
+    {
+      buffer_add_lexeme buffer lexbuf ;
+      if depth_curly = 0 then (
+        warning
+          input
+          ~start_offset:(Lexing.lexeme_end lexbuf)
+          (Parse_error.not_allowed
+            ~what:"'}' (end of reference)"
+            ~in_what:(
+              Printf.sprintf "'%s' (custom operator)"
+                (Buffer.sub buffer 0 ((Buffer.length buffer) - 1)))) ;
+        true )
+      else
+        ( reference_paren_content input start depth_paren (depth_curly - 1)
+            buffer lexbuf ) }
+  | eof
+    { false }
+  | _
+    {
+      buffer_add_lexeme buffer lexbuf ;
+      reference_paren_content input start depth_paren depth_curly buffer lexbuf }
+
+and reference_content input start buffer = parse
+  | '}'
+    {
+      Buffer.contents buffer
+    }
+  | '('
+    {
+      buffer_add_lexeme buffer lexbuf ;
+      let ref_closed = reference_paren_content input start 0 0 buffer lexbuf in
```

Comment:
Recursive calls are not the best way to call rules (eg. the double eof warning, the `}` character being added to the buffer).
I suggest tail-calling into `reference_paren_content` here and from there, tail-call back into `reference_content` only when needed. I think it's easier to follow that way, for example no need to wonder what happens when one function return but the caller continue to iterate.